Batches aren't usable in tokio tasks

Consider this example which spawns 10 tasks and, in each one of them, tries to execute a batch:

```rust
use anyhow::Result;
use scylla::batch::{Batch, BatchType};
use scylla::query::Query;
use scylla::transport::session::Session;
use scylla::SessionBuilder;
use std::env;
use std::sync::Arc;

#[tokio::main]
async fn main() -> Result<()> {
    let uri = env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());

    println!("Connecting to {} ...", uri);

    let session: Session = SessionBuilder::new().known_node(uri).build().await?;
    let session = Arc::new(session);

    // Spawn 10 tasks to execute some batches in parallel
    let mut tasks = Vec::new();
    for pk in 0..10 {
        let session = session.clone();
        let cks = (0..10).into_iter().collect::<Vec<_>>();
        tasks.push(tokio::task::spawn(async move {
            let q = Query::new("INSERT INTO ks.t (pk, ck) VALUES (?, ?)");
            let mut batch = Batch::new(BatchType::Unlogged);
            let mut vals: Vec<(i64, &i64)> = Vec::with_capacity(cks.len());

            // Deliberately use .iter() here instead of .into_iter()
            for ck in cks.iter() {
                batch.append_statement(q.clone());
                vals.push((pk, ck));
            }
            session.batch(&batch, vals).await.unwrap();
        }));
    }

    for task in tasks {
        task.await.unwrap();
    }

    Ok(())
}

```

This works on v0.6.1, but on current main (23d799414bc2caa1db9e4b0b1d1b459ab3efbdb6) it fails with the following error:

```
error: implementation of `BatchValuesGatWorkaround` is not general enough
  --> examples/problematic_batches.rs:23:20
   |
23 |           tasks.push(tokio::task::spawn(async move {
   |  ____________________^
24 | |             let q = Query::new("INSERT INTO ks.t (pk, ck) VALUES (?, ?)");
25 | |             let mut batch = Batch::new(BatchType::Unlogged);
26 | |             let mut vals: Vec<(i64, &i64)> = Vec::with_capacity(cks.len());
...  |
33 | |             session.batch(&batch, vals).await.unwrap();
34 | |         }));
   | |__________^ implementation of `BatchValuesGatWorkaround` is not general enough
   |
   = note: `Vec<(i64, &i64)>` must implement `BatchValuesGatWorkaround<'0, &'1 Vec<(i64, &i64)>>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `BatchValuesGatWorkaround<'2>`, for some specific lifetime `'2`

error: could not compile `examples` due to previous error
```

Applying the PR which switches to real GATs (https://github.com/scylladb/scylla-rust-driver/pull/596) just changes the error:

```
error: implementation of `Value` is not general enough
  --> examples/problematic_batches.rs:23:20
   |
23 |           tasks.push(tokio::task::spawn(async move {
   |  ____________________^
24 | |             let q = Query::new("INSERT INTO ks.t (pk, ck) VALUES (?, ?)");
25 | |             let mut batch = Batch::new(BatchType::Unlogged);
26 | |             let mut vals: Vec<(i64, &i64)> = Vec::with_capacity(cks.len());
...  |
33 | |             session.batch(&batch, vals).await.unwrap();
34 | |         }));
   | |__________^ implementation of `Value` is not general enough
   |
   = note: `Value` would have to be implemented for the type `&'0 i64`, for any lifetime `'0`...
   = note: ...but `Value` is actually implemented for the type `&'1 i64`, for some specific lifetime `'1`

error: could not compile `examples` due to previous error
```

If I change the type of `values` to be `Vec<(i64, i64)>` - so, no borrowing - then it works with the "real GATs" PR, but _still_ fails on main, with a similar error to the first one:

```
error: implementation of `BatchValuesGatWorkaround` is not general enough
  --> examples/problematic_batches.rs:23:20
   |
23 |           tasks.push(tokio::task::spawn(async move {
   |  ____________________^
24 | |             let q = Query::new("INSERT INTO ks.t (pk, ck) VALUES (?, ?)");
25 | |             let mut batch = Batch::new(BatchType::Unlogged);
26 | |             let mut vals: Vec<(i64, i64)> = Vec::with_capacity(cks.len());
...  |
33 | |             session.batch(&batch, vals).await.unwrap();
34 | |         }));
   | |__________^ implementation of `BatchValuesGatWorkaround` is not general enough
   |
   = note: `Vec<(i64, i64)>` must implement `BatchValuesGatWorkaround<'0, &'1 Vec<(i64, i64)>>`, for any two lifetimes `'0` and `'1`...
   = note: ...but it actually implements `BatchValuesGatWorkaround<'2>`, for some specific lifetime `'2`

error: could not compile `examples` due to previous error
```

A gat workaround version mentions "current limitations in the borrow checker"... :(

(With this additional commit compared to main: https://github.com/scylladb/scylla-rust-driver/commit/421005d8f9e61cb778294ca7fb14c53d89c3ea18)
```rust
use {futures::prelude::*, scylla::frame::value::BatchValues};

fn do_stuff_with_batch_values<'a>(
	values: Vec<(i64, &'a i64)>,
) -> std::pin::Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
	Box::pin(do_stuff_with_batch_values_async(values))
}

async fn do_stuff_with_batch_values_async<BV>(values: BV)
where
	BV: BatchValues,
	for<'r> <BV as scylla::frame::value::BatchValuesGatWorkaround<'r>>::BatchValuesIter: 'r + Send + Sync,
{
	let _iter = values.batch_values_iter();
	future::ready(()).await;
	// Iter drops here, across await boundary, so needs to be Send
}
```

```
error: implementation of `BatchValuesGatWorkaround` is not general enough
 --> src/lib.rs:6:2
  |
6 |     Box::pin(do_stuff_with_batch_values_async(values))
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `BatchValuesGatWorkaround` is not general enough
  |
  = note: `Vec<(i64, &i64)>` must implement `BatchValuesGatWorkaround<'0, &'1 Vec<(i64, &i64)>>`, for any two lifetimes `'0` and `'1`...
  = note: ...but it actually implements `BatchValuesGatWorkaround<'2>`, for some specific lifetime `'2`

error[E0521]: borrowed data escapes outside of function
  --> src/lib.rs:6:11
   |
3  | fn do_stuff_with_batch_values<'a>(
   |                               -- lifetime `'a` defined here
4  |     values: Vec<(i64, &'a i64)>,
   |     ------ `values` is a reference that is only valid in the function body
5  | ) -> std::pin::Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
6  |     Box::pin(do_stuff_with_batch_values_async(values))
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |              |
   |              `values` escapes the function body here
   |              argument requires that `'a` must outlive `'static`
   |
note: due to current limitations in the borrow checker, this implies a `'static` lifetime
  --> src/lib.rs:12:87
   |
12 |     for<'r> <BV as scylla::frame::value::BatchValuesGatWorkaround<'r>>::BatchValuesIter: 'r + Send + Sync,
   |                                                                                          ^^

For more information about this error, try `rustc --explain E0521`.
```

--- a/scylla-cql/src/frame/value.rs
+++ b/scylla-cql/src/frame/value.rs
@@ -250,7 +250,7 @@ pub trait BatchValuesIterator<'a> {
     fn skip_next(&mut self) -> Option<()>;
 }
 
-/// Implements `BatchValuesIterator` from an `Iterator` over things that implement `ValueList`
+/// Implements `BatchValuesIterator` from an `Iterator` over references to things that implement `ValueList`
 ///
 /// Essentially used internally by this lib to provide implementors of `BatchValuesIterator` for cases
 /// that always serialize the same concrete `ValueList` type

--- a/scylla-cql/src/frame/value.rs
+++ b/scylla-cql/src/frame/value.rs
@@ -223,6 +223,10 @@ pub trait BatchValues: for<'r> BatchValuesGatWorkaround<'r> {}
 impl<T: for<'r> BatchValuesGatWorkaround<'r> + ?Sized> BatchValues for T {}
 
 pub trait BatchValuesGatWorkaround<'r, ImplicitBounds = &'r Self> {
+    /// For some unknown reason, this type, when not resolved to a concrete type for a given async function,
+    /// cannot live across await boundaries while maintaining the corresponding future `Send`, unless `'r: 'static`
+    ///
+    /// See <https://github.com/scylladb/scylla-rust-driver/issues/599> for more details
     type BatchValuesIter: BatchValuesIterator<'r>;
     fn batch_values_iter(&'r self) -> Self::BatchValuesIter;
 }

--- a/scylla/src/transport/session.rs
+++ b/scylla/src/transport/session.rs
@@ -1030,10 +1030,7 @@ impl Session {
         // If users batch statements by shard, they will be rewarded with full shard awareness
 
         // Extract first serialized_value
-        let mut batch_values_iter_for_first_serialized_value = values.batch_values_iter();
-        let first_serialized_value = batch_values_iter_for_first_serialized_value
-            .next_serialized()
-            .transpose()?;
+        let first_serialized_value = values.batch_values_iter().next_serialized().transpose()?;
         let first_serialized_value = first_serialized_value.as_deref();
         let statement_info = match (first_serialized_value, batch.statements.first()) {
             (Some(first_serialized_value), Some(BatchStatement::PreparedStatement(ps))) => {

--- a/scylla/src/transport/session_test.rs
+++ b/scylla/src/transport/session_test.rs
@@ -22,6 +22,7 @@ use itertools::Itertools;
 use scylla_cql::frame::value::Value;
 use std::collections::BTreeSet;
 use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
 use tokio::net::TcpListener;
 use uuid::Uuid;
 
@@ -358,7 +359,7 @@ async fn test_prepared_statement() {
 #[tokio::test]
 async fn test_batch() {
     let uri = std::env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());
-    let session = SessionBuilder::new().known_node(uri).build().await.unwrap();
+    let session = Arc::new(SessionBuilder::new().known_node(uri).build().await.unwrap());
     let ks = unique_keyspace_name();
 
     session.query(format!("CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class' : 'SimpleStrategy', 'replication_factor' : 1}}", ks), &[]).await.unwrap();
@@ -390,9 +391,23 @@ async fn test_batch() {
     batch.append_statement(prepared_statement.clone());
 
     let four_value: i32 = 4;
-    let values = ((1_i32, 2_i32, "abc"), (), (1_i32, &four_value, "hello"));
-
-    session.batch(&batch, values).await.unwrap();
+    let hello_value: String = String::from("hello");
+    let session_clone = session.clone();
+    // We're spawning to a separate task here to test that it works even in that case, because in some scenarios
+    // (specifically if the `BatchValuesIter` associated type is not dropped before await boundaries)
+    // the implicit auto trait propagation on batch will be such that the returned future is not Send (depending on
+    // some lifetime for some unknown reason), so can't be spawned on tokio.
+    // See https://github.com/scylladb/scylla-rust-driver/issues/599 for more details
+    tokio::spawn(async move {
+        let values = (
+            (1_i32, 2_i32, "abc"),
+            (),
+            (1_i32, &four_value, hello_value.as_str()),
+        );
+        session_clone.batch(&batch, values).await.unwrap();
+    })
+    .await
+    .unwrap();
 
     let rs = session
         .query(format!("SELECT a, b, c FROM {}.t_batch", ks), &[])

